tuf/tuf.go

```diff
@@ -173,6 +173,112 @@ func (tr *Repo) RemoveBaseKeys(role string, keyIDs ...string) error {
 	return nil
 }
 
+// GetBaseRole gets a base role from this repo's metadata
+func (tr *Repo) GetBaseRole(name string) (data.BaseRole, error) {
+	if !data.IsBaseRole(name) {
+		return data.BaseRole{}, data.ErrInvalidRole{Role: name, Reason: "invalid base role name"}
+	}
+	if tr.Root == nil {
+		return data.BaseRole{}, ErrNotLoaded{data.CanonicalRootRole}
+	}
+	roleData, ok := tr.Root.Signed.Roles[name]
+	if !ok {
+		return data.BaseRole{}, ErrNotLoaded{name}
+	}
+	// Get all public keys for the base role from TUF metadata
+	keyIDs := roleData.KeyIDs
+	pubKeys := make(map[string]data.PublicKey)
+	for _, keyID := range keyIDs {
+		pubKey, ok := tr.Root.Signed.Keys[keyID]
+		if !ok {
+			return data.BaseRole{}, fmt.Errorf("key with ID %s, specified as a signing key for role %s, was not found in root metadata", keyID, name)
+		}
+		pubKeys[keyID] = pubKey
+	}
+
+	return data.BaseRole{
+		Name:      name,
+		Keys:      pubKeys,
+		Threshold: roleData.Threshold,
+	}, nil
+}
+
+// GetDelegationRole gets a delegation role from this repo's metadata, walking from the targets role down to the delegation itself
+func (tr *Repo) GetDelegationRole(name string) (data.DelegationRole, error) {
+	if !data.IsDelegation(name) {
+		return data.DelegationRole{}, data.ErrInvalidRole{Role: name, Reason: "invalid delegation name"}
+	}
+	if tr.Root == nil {
+		return data.DelegationRole{}, ErrNotLoaded{data.CanonicalRootRole}
+	}
+	_, ok := tr.Root.Signed.Roles[data.CanonicalTargetsRole]
+	if !ok {
+		return data.DelegationRole{}, ErrNotLoaded{data.CanonicalTargetsRole}
+	}
+	// Traverse target metadata, down to delegation itself
+	// Get all public keys for the base role from TUF metadata
+	signedTargetData, ok := tr.Targets[data.CanonicalTargetsRole]
+	if !ok {
+		return data.DelegationRole{}, fmt.Errorf("TUF data for targets does not exist in targets metadata", name)
+	}
+
+	// Keep track of paths and path hash prefixes to restrict as we traverse the delegations tree, targets implicitly has ""
+	whiteListedPaths := []string{""}
```

Comment:
It looks like we are pruning paths that shouldn't exist, as we go?  This means that the in-memory data structure contain only the valid paths for the role, which makes sense if we want to list targets, etc.  But would this mean that if a target was added, or any other changes were made, to the role, then a possibly unintended paths/paths-prefix change would be published when the user publishes whatever their change is?  Since when we publish we re-serialize this in-memory structure.  Is that behavior that we want?